File: dashboard/src/competitorApi.js
/**
 * Client for the competitor-study API.
 *
 * One place that knows the endpoint shapes and how this backend reports errors:
 * failures come back as `{ error, detail }` with a non-2xx status, so every call
 * funnels through `request()` and throws an Error carrying the server's message.
 * Components then only handle `try/catch`, never response plumbing.
 */

const BASE = '/api/competitor';

async function request(path, { method = 'GET', body, signal } = {}) {
  const response = await fetch(`${BASE}${path}`, {
    method,
    credentials: 'include',
    headers: body ? { 'Content-Type': 'application/json' } : undefined,
    body: body ? JSON.stringify(body) : undefined,
    signal,
  });

  let payload;
  try {
    payload = await response.json();
  } catch {
    payload = null;   // empty or non-JSON body (e.g. 204, or an HTML error page)
  }

  if (!response.ok) {
    const message =
      payload?.detail || payload?.error || `Request failed (${response.status})`;
    const error = new Error(message);
    error.status = response.status;
    throw error;
  }
  return payload ?? {};
}

// --- studies ---------------------------------------------------------------
export const listStudies = () => request('/studies');
export const createStudy = (body) => request('/studies', { method: 'POST', body });
export const getStudy = (id) => request(`/studies/${id}`);
export const updateStudy = (id, body) => request(`/studies/${id}`, { method: 'PUT', body });
export const deleteStudy = (id) => request(`/studies/${id}`, { method: 'DELETE' });
// --- business profile ------------------------------------------------------
export const getProfile = (id) => request(`/studies/${id}/profile`);
/** Scrapes the website and derives market context — expect this to take a while. */
export const buildProfile = (id, body) => request(`/studies/${id}/profile`, { method: 'POST', body });
export const saveProfile = (id, body) => request(`/studies/${id}/profile`, { method: 'PUT', body });

// --- competitors -----------------------------------------------------------
/** Queues discovery as a background job; returns { run_id, status } immediately.
 *  Poll getDiscoveryStatus(id, run_id) until status is 'success' or 'failed'. */
export const discoverCompetitors = (id, body) => request(`/studies/${id}/discover`, { method: 'POST', body });
export const getDiscoveryStatus = (id, runId) => request(`/studies/${id}/discover/${runId}`);
/** Phase 3: finds channels for every tracked competitor that doesn't have one yet.
 *  Same queued/poll shape as discoverCompetitors — reuse pollDiscoveryRun on the result. */
export const discoverTrackedAccounts = (id) => request(`/studies/${id}/discover-accounts`, { method: 'POST' });
export const listCompetitors = (id) => request(`/studies/${id}/competitors`);
export const addCompetitor = (id, body) => request(`/studies/${id}/competitors`, { method: 'POST', body });
/** Creates a competitor and validates+links its sources in one call — the
 *  manual-first path, as opposed to addCompetitor()+addAccount() one at a time. */
export const addCompetitorManual = (id, body) =>
  request(`/studies/${id}/competitors/manual`, { method: 'POST', body });
export const setCompetitorStatus = (competitorId, status) =>
  request(`/competitors/${competitorId}/status`, { method: 'POST', body: { status } });
/** Partial update - the backend merges onto the stored row, so a body of just
 *  { aliases } leaves every other field alone. */
export const updateCompetitor = (competitorId, body) =>
  request(`/competitors/${competitorId}`, { method: 'PUT', body });
export const deleteCompetitor = (competitorId) =>
  request(`/competitors/${competitorId}`, { method: 'DELETE' });

// --- accounts --------------------------------------------------------------
export const listAccounts = (competitorId) => request(`/competitors/${competitorId}/accounts`);
export const discoverAccounts = (competitorId) =>
  request(`/competitors/${competitorId}/accounts/discover`, { method: 'POST' });
export const addAccount = (competitorId, body) =>
  request(`/competitors/${competitorId}/accounts`, { method: 'POST', body });
export const validateAccount = (accountId, status, reason = '') =>
  request(`/accounts/${accountId}/validate`, { method: 'POST', body: { status, reason } });
export const deleteAccount = (accountId) => request(`/accounts/${accountId}`, { method: 'DELETE' });

// --- scheduling ------------------------------------------------------------
export const getSchedule = (id) => request(`/studies/${id}/schedule`);
export const setSchedule = (id, body) => request(`/studies/${id}/schedule`, { method: 'PUT', body });
export const syncSources = (id) => request(`/studies/${id}/sync-sources`, { method: 'POST' });

// --- shared presentation helpers ------------------------------------------
export const SIZE_TIER_LABELS = {
  enterprise: 'Enterprise',
  mid_market: 'Mid-market',
  smb: 'SMB',
  startup: 'Startup',
  unknown: 'Unknown size',
};

export const SOURCE_KIND_OPTIONS = [
  { value: 'rss', label: 'RSS' },
  { value: 'web', label: 'Web' },
  { value: 'social', label: 'Social' },
  { value: 'hashtag', label: 'Hashtag' },
  { value: 'keyword', label: 'Keyword' },
  { value: 'username', label: 'X Account' },
  { value: 'reddit', label: 'Reddit' },
  { value: 'telegram', label: 'Telegram' },
];

/** These take a bare name/term instead of a URL — the source row hides its
 *  URL field and derives the real URL server-side. */
export const TERM_SOURCE_TYPES = new Set(['hashtag', 'keyword', 'username']);

export const TERM_SOURCE_PLACEHOLDERS = {
  hashtag: 'Hashtag, without # (e.g. EVSummit)',
  username: 'X account, without @ (e.g. elonmusk)',
  keyword: 'Keyword or phrase (e.g. electric vehicles)',
};

export const PLATFORM_LABELS = Object.fromEntries(
  SOURCE_KIND_OPTIONS.map((option) => [option.value, option.label]),
);

/** Shape-only check mirroring the backend's normalize_source_url — enough to
 *  catch an obviously-broken entry before it round-trips to the server. */
export function isPlausibleUrl(value) {
  const text = String(value || '').trim();
  if (!text) return false;
  const withScheme = /^https?:\/\//i.test(text) ? text : `https://${text}`;
  try {
    const { hostname } = new URL(withScheme);
    return hostname.includes('.');
  } catch {
    return false;
  }
}

const DISCOVERY_POLL_MS = 2500;

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Discovery runs as a backend job that can take minutes (LLM call + live web
 *  corroboration + per-competitor account lookups), so it's queued rather than
 *  awaited directly - poll until it reaches a terminal status. Shared by
 *  onboarding and the workspace, which both trigger discovery.
 *  `onUpdate`, if given, is called with the run on every poll (not just the
 *  terminal one) so a caller can render its live `run.logs` as they arrive. */
export async function pollDiscoveryRun(studyId, runId, onUpdate) {
  for (;;) {
    const { run } = await getDiscoveryStatus(studyId, runId);
    if (onUpdate) onUpdate(run);
    if (run.status === 'success' || run.status === 'failed') return run;
    await sleep(DISCOVERY_POLL_MS);
  }
}

/** Initials for a competitor avatar, e.g. "Blue Yonder" -> "BY". */
export function initials(name) {
  const words = String(name || '')
    .trim()
    .split(/\s+/)
    .filter(Boolean);
  if (!words.length) return '?';
  if (words.length === 1) return words[0].slice(0, 2).toUpperCase();
  return (words[0][0] + words[1][0]).toUpperCase();
}

/** Stable per-competitor avatar colour, so a company looks the same everywhere. */
export function avatarGradient(name) {
  const palette = [
    ['#6366f1', '#4338ca'],
    ['#0ea5e9', '#0369a1'],
    ['#14b8a6', '#0f766e'],
    ['#f59e0b', '#b45309'],
    ['#ec4899', '#be185d'],
    ['#8b5cf6', '#6d28d9'],
  ];
  let hash = 0;
  const text = String(name || '');
  for (let i = 0; i < text.length; i += 1) {
    hash = (hash * 31 + text.charCodeAt(i)) % 997;
  }
  const [from, to] = palette[hash % palette.length];
  return `linear-gradient(135deg, ${from}, ${to})`;
}

export function relativeTime(value) {
  if (!value) return null;
  const diff = Date.now() - new Date(value).getTime();
  if (!Number.isFinite(diff)) return null;
  if (diff < 60_000) return 'just now';
  const minutes = Math.floor(diff / 60_000);
  if (minutes < 60) return `${minutes}m ago`;
  const hours = Math.floor(minutes / 60);
  if (hours < 24) return `${hours}h ago`;
  const days = Math.floor(hours / 24);
  if (days < 30) return `${days}d ago`;
  return new Date(value).toLocaleDateString();
}

export function formatDate(value) {
  if (!value) return null;
  const date = new Date(value);
  if (Number.isNaN(date.getTime())) return null;
  return date.toLocaleDateString(undefined, { year: 'numeric', month: 'short', day: 'numeric' });
}

File: backend/services/competitors/competitor_api.py
"""HTTP surface for the competitor study.

Kept in its own APIRouter rather than appended to main.py: this is a separate
experience from opinion monitoring, and separating the routes is what keeps
the two from tangling as either one grows.

Collection only - a study defines *who* to watch and which channels to
collect from; it draws no conclusions. Findings generation lives in whatever
analyzes the exported articles.

Long-running work (the business-website scrape) runs synchronously and is
expected to take tens of seconds — the UI shows staged progress for it,
matching how the existing project discovery flow behaves. Competitor
discovery is the exception: it runs as a background job (see
discover()/discover_status() below) because it can take minutes once web
corroboration and per-competitor account lookups are added up, well past any
gateway timeout.
"""

from __future__ import annotations


from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException

from services.competitors import business_profile_store
from services.competitors import competitor_discovery
from services.competitors import competitors_store
from services.competitors import cultural_analysis_store
from services.competitors.countries import validate_countries
from services.projects.projects_store import REPEAT_WEEKDAYS
from psycopg.types.json import Jsonb
import db
from services.auth.auth import require_permission
from services.pipeline.pipeline_runs import get_active_run_for_project
from services.projects.projects_store import delete_project, list_sources_for_project, project_has_articles

router = APIRouter(prefix="/api/competitor", tags=["competitor"])


def _project_or_404(project_id: int) -> dict:
    project = db.fetch_one(
        "select id, name, mode, status, repeat_enabled, next_run_at, last_run_at, last_run_status "
        "from projects where id = %s",
        (int(project_id),),
    )
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return project


def _competitor_or_404(competitor_id: int) -> dict:
    competitor = competitors_store.get_competitor(competitor_id)
    if not competitor:
        raise HTTPException(status_code=404, detail="Competitor not found")
    return competitor


# --------------------------------------------------------------------------- #
# Studies
# --------------------------------------------------------------------------- #
@router.get("/studies")
def list_studies(user: dict = Depends(require_permission("competitors.view"))):
    """Competitor-mode projects with enough summary to render the index."""
    return {
        "studies": db.fetch_all(
            """
            with latest_findings as (
                -- One row per competitor's *current* card, not per generation
                -- event: generate_finding() always inserts (never updates), so
                -- re-running analysis on the same competitor leaves its older
                -- findings in place as history rather than superseding them in
                -- place. Counting competitor_findings directly counted every
                -- one of those, so the number grew on every re-run even with
                -- the competitor set unchanged, and didn't match the study's
                -- own findings grid - which shows one card per competitor, the
                -- newest, excluding rejected ones.
                select distinct on (competitor_id)
                       project_id, competitor_id, impact_level, generated_at
                from competitor_findings
                where validation_status != 'rejected'
                order by competitor_id, generated_at desc
            )
            select p.id, p.name, p.status, p.mode, p.created_at, p.updated_at,
                   p.repeat_enabled, p.next_run_at, p.last_run_at, p.last_run_status,
                   bp.name as business_name, bp.website as business_website,
                   bp.market, bp.industry, bp.scrape_status,
                   coalesce(c.tracked, 0)::int   as tracked_competitors,
                   coalesce(c.suggested, 0)::int as suggested_competitors,
                   coalesce(a.total, 0)::int     as article_count,
                   a.last_scraped_at
            from projects p
            left join business_profiles bp on bp.project_id = p.id
            left join (
                select project_id,
                       count(*) filter (where status = 'tracked')   as tracked,
                       count(*) filter (where status = 'suggested') as suggested
                from competitors group by project_id
            ) c on c.project_id = p.id
            left join (
                select ap.project_id,
                       count(*) as total,
                       max(ar.fetched_at) as last_scraped_at
                from article_projects ap
                join articles ar on ar.id = ap.article_id
                group by ap.project_id
            ) a on a.project_id = p.id
            where p.mode = 'competitor'
            order by p.created_at desc
            """
        )
    }


@router.post("/studies")
def create_study(payload: dict, user: dict = Depends(require_permission("competitors.manage"))):
    """Create a competitor-mode project. The business profile is added next."""
    name = str((payload or {}).get("name") or "").strip()
    if not name:
        raise HTTPException(status_code=400, detail="A study name is required.")
    project = db.fetch_one(
        """
        insert into projects (name, mode, status, description)
        values (%s, 'competitor', %s, %s)
        returning id, name, mode, status, created_at
        """,
        (name, str((payload or {}).get("status") or "active"),
         str((payload or {}).get("description") or "").strip() or None),
    )
    if not project:
        raise HTTPException(status_code=500, detail="Could not create the study.")
    return {"study": project}


@router.get("/studies/{project_id}")
def get_study(project_id: int, user: dict = Depends(require_permission("competitors.view"))):
    project = _project_or_404(project_id)
    return {
        "study": project,
        "profile": business_profile_store.get_profile(project_id),
        "competitors": competitors_store.competitor_overview(project_id),
        "cultural_analysis": cultural_analysis_store.get_analysis(project_id),
    }


@router.put("/studies/{project_id}")
def update_study(project_id: int, payload: dict, user: dict = Depends(require_permission("competitors.manage"))):
    _project_or_404(project_id)
    payload = payload or {}
    name = str(payload.get("name") or "").strip()
    if not name:
        raise HTTPException(status_code=400, detail="A study name is required.")
    project = db.fetch_one(
        """
        update projects
           set name = %s, status = %s, description = %s, updated_at = now()
         where id = %s and mode = 'competitor'
        returning id, name, mode, status, description, created_at, updated_at
        """,
        (name, str(payload.get("status") or "active"),
         str(payload.get("description") or "").strip() or None, int(project_id)),
    )
    if not project:
        raise HTTPException(status_code=404, detail="Study not found")
    return {"study": project}


@router.delete("/studies/{project_id}")
def remove_study(project_id: int, user: dict = Depends(require_permission("competitors.manage"))):
    _project_or_404(project_id)
    if not delete_project(project_id):
        raise HTTPException(status_code=500, detail="Unable to delete the study.")
    return {"ok": True}


# --------------------------------------------------------------------------- #
# Business profile
# --------------------------------------------------------------------------- #
@router.get("/studies/{project_id}/profile")
def get_profile(project_id: int, user: dict = Depends(require_permission("competitors.view"))):
    _project_or_404(project_id)
    return {"profile": business_profile_store.get_profile(project_id)}


@router.post("/studies/{project_id}/profile")
def build_profile(project_id: int, payload: dict, user: dict = Depends(require_permission("competitors.manage"))):
    """Scrape the business's website and derive its market context.

    Returns the scrape outcome alongside the profile so the UI can say how many
    pages were actually read — a thin scrape produces a weak profile, and that
    should be visible rather than inferred later from poor competitor matches.
    """
    _project_or_404(project_id)
    payload = payload or {}
    if not str(payload.get("name") or "").strip():
        raise HTTPException(status_code=400, detail="A business name is required.")
    if payload.get("target_countries") is not None and not isinstance(payload["target_countries"], list):
        raise HTTPException(status_code=400, detail="target_countries must be a list of ISO country codes.")
    return business_profile_store.build_profile(project_id, payload)


@router.put("/studies/{project_id}/profile")
def update_profile(project_id: int, payload: dict, user: dict = Depends(require_permission("competitors.manage"))):
    """Save user edits to the profile without re-scraping or re-deriving."""
    _project_or_404(project_id)
    payload = payload or {}
    if payload.get("target_countries") is not None and not isinstance(payload["target_countries"], list):
        raise HTTPException(status_code=400, detail="target_countries must be a list of ISO country codes.")
    existing = business_profile_store.get_profile(project_id) or {}
    merged = {**existing, **payload}
    profile = business_profile_store.upsert_profile(project_id, merged)
    if not profile:
        raise HTTPException(status_code=400, detail="Could not save the profile.")
    return {"profile": profile}


# --------------------------------------------------------------------------- #
# Cultural analysis
# --------------------------------------------------------------------------- #
@router.get("/studies/{project_id}/cultural-analysis")
def get_cultural_analysis(project_id: int, user: dict = Depends(require_permission("competitors.view"))):
    _project_or_404(project_id)
    return {"cultural_analysis": cultural_analysis_store.get_analysis(project_id)}


@router.post("/studies/{project_id}/cultural-analysis")
def run_cultural_analysis(project_id: int, user: dict = Depends(require_permission("competitors.analyze"))):
    """Assess how well the business fits the culture(s) it's targeting.

    Requires a business profile with target countries already set — there is
    nothing region-specific to analyze otherwise.
    """
    _project_or_404(project_id)
    try:
        return {"cultural_analysis": cultural_analysis_store.build_analysis(project_id)}
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))


# --------------------------------------------------------------------------- #
# Discovery
# --------------------------------------------------------------------------- #
@router.post("/studies/{project_id}/discover")
def discover(
    project_id: int,
    background_tasks: BackgroundTasks,
    payload: dict = None,
    user: dict = Depends(require_permission("competitors.analyze")),
):
    """Queue competitor discovery as a background job and return immediately.

    Discovery chains an LLM call, live web corroboration per candidate, and
    (with_accounts) a further LLM call per competitor - easily minutes end to
    end, which running inline used to push past the gateway timeout and 504.
    The UI polls GET .../discover/{run_id} for progress and refetches
    competitors once the run succeeds.
    """
    _project_or_404(project_id)
    profile = business_profile_store.get_profile(project_id)
    if not profile:
        raise HTTPException(status_code=400, detail="Add the business profile before discovering competitors.")

    payload = payload or {}
    limit = max(3, min(int(payload.get("limit") or competitor_discovery.MAX_COMPETITORS), 20))
    with_accounts = bool(payload.get("with_accounts"))

    active = competitor_discovery.get_active_discovery_run(project_id)
    if active:
        return {"run_id": active["run_id"], "status": active["status"]}

    run_id = competitor_discovery.create_discovery_run(project_id)
    background_tasks.add_task(
        competitor_discovery.run_discovery_job, run_id, project_id, profile, limit, with_accounts
    )
    return {"run_id": run_id, "status": "queued", "model": competitor_discovery.discovery_model()}


@router.get("/studies/{project_id}/discover/{run_id}")
def discover_status(project_id: int, run_id: str, user: dict = Depends(require_permission("competitors.view"))):
    _project_or_404(project_id)
    run = competitor_discovery.get_discovery_run(run_id)
    if not run or run["project_id"] != project_id:
        raise HTTPException(status_code=404, detail="Discovery run not found.")
    return {"run": run}


@router.post("/studies/{project_id}/discover-accounts")
def discover_accounts_bulk(
    project_id: int,
    background_tasks: BackgroundTasks,
    user: dict = Depends(require_permission("competitors.analyze")),
):
    """Phase 3: find channels for every tracked competitor that doesn't have one yet.

    Deliberately separate from discover() (Phase 1, finding the competitors
    themselves) so channel discovery only spends LLM calls on companies the user
    actually chose to track, not every AI suggestion.
    """
    _project_or_404(project_id)
    targets = [
        {"id": c["id"], "name": c["name"], "website": c.get("website")}
        for c in competitors_store.competitor_overview(project_id)
        if c["status"] == "tracked" and not c.get("account_count")
    ]
    if not targets:
        return {"run_id": None, "status": "success", "message": "No tracked competitors need channels."}

    active = competitor_discovery.get_active_discovery_run(project_id)
    if active:
        return {"run_id": active["run_id"], "status": active["status"]}

    run_id = competitor_discovery.create_discovery_run(project_id)
    background_tasks.add_task(competitor_discovery.run_accounts_discovery_job, run_id, project_id, targets)
    return {"run_id": run_id, "status": "queued"}


@router.post("/competitors/{competitor_id}/accounts/discover")
def discover_competitor_accounts(competitor_id: int, user: dict = Depends(require_permission("competitors.analyze"))):
    competitor = _competitor_or_404(competitor_id)
    profile = business_profile_store.get_profile(competitor["project_id"]) or {}
    target_countries = validate_countries(profile.get("target_countries"))
    found = competitor_discovery.discover_accounts(competitor["name"], competitor.get("website"), target_countries)
    for account in found:
        competitors_store.upsert_account(competitor_id, account)
    return {"accounts": competitors_store.list_accounts(competitor_id)}


# --------------------------------------------------------------------------- #
# Competitors + accounts
# --------------------------------------------------------------------------- #
@router.get("/studies/{project_id}/competitors")
def list_competitors(project_id: int, user: dict = Depends(require_permission("competitors.view"))):
    _project_or_404(project_id)
    competitors = competitors_store.competitor_overview(project_id)
    for competitor in competitors:
        competitor["accounts"] = competitors_store.list_accounts(competitor["id"])
    return {"competitors": competitors}


@router.post("/studies/{project_id}/competitors")
def add_competitor(project_id: int, payload: dict, user: dict = Depends(require_permission("competitors.manage"))):
    _project_or_404(project_id)
    payload = payload or {}
    if not str(payload.get("name") or "").strip():
        raise HTTPException(status_code=400, detail="A competitor name is required.")
    record = competitors_store.upsert_competitor(
        project_id, {**payload, "discovery_source": payload.get("discovery_source") or "manual"}
    )
    if not record:
        raise HTTPException(status_code=400, detail="Could not save the competitor.")
    competitors_store.rerank_competitors(project_id)
    return {"competitor": record}


@router.post("/studies/{project_id}/competitors/manual")
def add_competitor_manual(project_id: int, payload: dict, user: dict = Depends(require_permission("competitors.manage"))):
    """Create (or merge into) a competitor and validate its sources immediately.

    Manual sources skip the pending/confirm step AI-discovered accounts need:
    the user typed the URL themselves, so there is no attribution risk to guard
    against. Every source is validated before anything is written, so a bad
    URL fails the whole request rather than leaving a half-created competitor.
    """
    _project_or_404(project_id)
    payload = payload or {}
    name = str(payload.get("name") or "").strip()
    if not name:
        raise HTTPException(status_code=400, detail="A competitor name is required.")

    raw_sources = payload.get("sources") or []
    if not isinstance(raw_sources, list):
        raise HTTPException(status_code=400, detail="sources must be a list.")

    normalized_sources = []
    for index, source in enumerate(raw_sources, start=1):
        source = source or {}
        platform = str(source.get("platform") or "").strip().lower()
        if platform not in competitors_store.PLATFORM_SOURCE_TYPE:
            raise HTTPException(status_code=400, detail=f"Source {index}: unsupported source type '{platform}'.")
        handle_input = str(source.get("handle") or "").strip().lstrip("@")
        url = competitors_store.resolve_account_url(platform, source.get("url"), handle_input)
        if not url:
            raise HTTPException(status_code=400, detail=f"Source {index}: enter a valid value.")
        normalized_sources.append({
            "platform": platform,
            "url": url,
            "handle": handle_input or None,
        })

    competitor = competitors_store.upsert_competitor(
        project_id, {**payload, "status": "tracked", "discovery_source": "manual"}
    )
    if not competitor:
        raise HTTPException(status_code=400, detail="Could not save the competitor.")

    accounts = []
    for source in normalized_sources:
        account = competitors_store.upsert_account(
            competitor["id"], {**source, "validation_status": "valid", "confidence": 1.0}
        )
        if account:
            accounts.append(account)

    competitors_store.rerank_competitors(project_id)
    return {"competitor": competitor, "accounts": accounts}


@router.put("/competitors/{competitor_id}")
def update_competitor(competitor_id: int, payload: dict, user: dict = Depends(require_permission("competitors.manage"))):
    competitor = _competitor_or_404(competitor_id)
    record = competitors_store.upsert_competitor(
        competitor["project_id"], {**competitor, **(payload or {})}
    )
    competitors_store.rerank_competitors(competitor["project_id"])
    return {"competitor": record}


@router.post("/competitors/{competitor_id}/status")
def set_status(competitor_id: int, payload: dict, user: dict = Depends(require_permission("competitors.manage"))):
    """Track or ignore a competitor. Tracking links its valid accounts as sources.

    Phase 2: the first time an AI-suggested competitor is tracked, corroborate
    it against the live web here — the check discover() skips on every
    candidate up front now runs once, only on the one the user actually chose.
    Never blocks the track itself; a company that fails to corroborate still
    gets tracked, just flagged, since the user has already decided to track it.
    """
    competitor = _competitor_or_404(competitor_id)
    status = str((payload or {}).get("status") or "").strip().lower()
    record = competitors_store.set_competitor_status(competitor_id, status)
    if not record:
        raise HTTPException(status_code=400, detail="status must be suggested, tracked, or ignored.")

    verification = None
    signals = record.get("size_signals") or {}
    if status == "tracked" and record.get("discovery_source") == "ai" and not signals.get("checked_live"):
        verification = competitor_discovery.verify_competitor(record["name"], record.get("website"))
        record = competitors_store.upsert_competitor(competitor["project_id"], {
            **record,
            "website": verification["resolved_website"] or record.get("website"),
            "size_signals": {
                **signals,
                "checked_live": True,
                "site_reachable": verification["reachable"],
                "search_hits": verification["search_hits"],
            },
        }) or record

    sync = competitors_store.sync_project_sources(competitor["project_id"])
    return {"competitor": record, "sources": sync, "verification": verification}


@router.delete("/competitors/{competitor_id}")
def remove_competitor(competitor_id: int, user: dict = Depends(require_permission("competitors.manage"))):
    competitor = _competitor_or_404(competitor_id)
    competitors_store.delete_competitor(competitor_id)
    competitors_store.rerank_competitors(competitor["project_id"])
    return {"ok": True}


@router.get("/competitors/{competitor_id}/accounts")
def list_accounts(competitor_id: int, user: dict = Depends(require_permission("competitors.view"))):
    _competitor_or_404(competitor_id)
    return {"accounts": competitors_store.list_accounts(competitor_id)}


@router.post("/competitors/{competitor_id}/accounts")
def add_account(competitor_id: int, payload: dict, user: dict = Depends(require_permission("competitors.manage"))):
    _competitor_or_404(competitor_id)
    account = competitors_store.upsert_account(competitor_id, payload or {})
    if not account:
        raise HTTPException(status_code=400, detail="platform and url are required.")
    return {"account": account}


@router.post("/accounts/{account_id}/validate")
def validate_account(account_id: int, payload: dict, user: dict = Depends(require_permission("competitors.manage"))):
    """Accept or reject a discovered account.

    Accepting registers it as a scrape source; rejecting detaches it. Nothing is
    scraped on a guess, because a misattributed account puts another company's
    activity into a report someone plans against.
    """
    status = str((payload or {}).get("status") or "").strip().lower()
    account = competitors_store.set_account_validation(
        account_id, status, str((payload or {}).get("reason") or "")
    )
    if not account:
        raise HTTPException(status_code=400, detail="status must be pending, valid, or rejected.")
    return {"account": account}


@router.delete("/accounts/{account_id}")
def remove_account(account_id: int, user: dict = Depends(require_permission("competitors.manage"))):
    competitors_store.delete_account(account_id)
    return {"ok": True}


# --------------------------------------------------------------------------- #
# Scheduling — reuses the existing project scheduler, no new machinery
# --------------------------------------------------------------------------- #
@router.get("/studies/{project_id}/schedule")
def get_schedule(project_id: int, user: dict = Depends(require_permission("competitors.view"))):
    return {"schedule": db.fetch_one(
        """
        select repeat_enabled, repeat_interval_value, repeat_interval_unit,
               repeat_weekdays, first_run_at, next_run_at, last_run_at, last_run_status,
               start_date, end_date
        from projects where id = %s
        """,
        (int(project_id),),
    )}


@router.put("/studies/{project_id}/schedule")
def set_schedule(project_id: int, payload: dict, user: dict = Depends(require_permission("competitors.manage"))):
    """Enable recurring competitor scrapes via the existing project scheduler.

    Also carries the data retrieval window (start_date/end_date) - the same
    columns Opinion Monitor projects use to scope which article publish dates
    get pulled in. The dashboard keeps this window's span in sync with the
    repeat interval; the backend just persists whatever it's given.
    """
    _project_or_404(project_id)
    payload = payload or {}
    enabled = bool(payload.get("repeat_enabled"))
    unit = str(payload.get("repeat_interval_unit") or "days").strip().lower()
    if unit not in {"minutes", "hours", "days"}:
        raise HTTPException(status_code=400, detail="repeat_interval_unit must be minutes, hours, or days.")
    try:
        value = max(1, int(payload.get("repeat_interval_value") or 1))
    except (TypeError, ValueError):
        raise HTTPException(status_code=400, detail="repeat_interval_value must be a positive number.")

    first_run = payload.get("first_run_at") or None
    start_date = str(payload.get("start_date") or "").strip() or None
    end_date = str(payload.get("end_date") or "").strip() or None

    # Only touch weekdays when the caller actually sent the field - the
    # workspace's simpler reschedule call doesn't, and a bare interval change
    # there must not silently clear a day restriction set during onboarding.
    weekdays_param = None
    if "repeat_weekdays" in payload:
        seen = set()
        weekdays = []
        for day in payload.get("repeat_weekdays") or []:
            text = str(day or "").strip().lower()
            if text and text in REPEAT_WEEKDAYS and text not in seen:
                seen.add(text)
                weekdays.append(text)
        weekdays_param = Jsonb(weekdays)

    schedule = db.fetch_one(
        """
        update projects
           set repeat_enabled = %s,
               repeat_interval_value = %s,
               repeat_interval_unit = %s,
               first_run_at = coalesce(%s::timestamptz, first_run_at),
               next_run_at = case
                   when %s then coalesce(%s::timestamptz, next_run_at, now())
                   else null
               end,
               start_date = coalesce(%s::date, start_date),
               end_date = coalesce(%s::date, end_date),
               repeat_weekdays = coalesce(%s::jsonb, repeat_weekdays)
         where id = %s
        returning repeat_enabled, repeat_interval_value, repeat_interval_unit,
                  repeat_weekdays, first_run_at, next_run_at, last_run_at, last_run_status,
                  start_date, end_date
        """,
        (enabled, value, unit, first_run, enabled, first_run, start_date, end_date, weekdays_param, int(project_id)),
    )
    return {"schedule": schedule}
